src/Endpoint/Partition.php

```diff
@@ -0,0 +1,171 @@
+        }
+
+        if (isset($this->definition['regionRegex'])) {
+            return (bool) preg_match(
+                "@{$this->definition['regionRegex']}@",
+                $region
+            );
+        }
+
+        return false;
+    }
+
+    public function getRegionsForService($service)
+    {
+        if (isset($this->definition['services'][$service]['isRegionalized'])
+            && false === $this->definition['services'][$service]['isRegionalized']
+        ) {
+            return [$this->definition['services'][$service]['partitionEndpoint']];
+        }
+
+        if (isset($this->definition['services'][$service]['endpoints'])) {
+            $serviceRegions = array_keys(
+                $this->definition['services'][$service]['endpoints']
+            );
+
+            return array_intersect($serviceRegions, array_keys(
+                $this->definition['regions']
+            ));
+        }
+
+        return [];
+    }
+
+    public function __invoke(array $args = [])
+    {
+        $service = isset($args['service']) ? $args['service'] : '';
+        $region = isset($args['region']) ? $args['region'] : '';
+        $scheme = isset($args['scheme']) ? $args['scheme'] : 'https';
+        $data = $this->getEndpointData($service, $region);
+
+        return [
+            'endpoint' => "{$scheme}://" . $this->formatEndpoint(
+                    isset($data['hostname']) ? $data['hostname'] : '',
+                    $service,
+                    $region
+                ),
+            'signatureVersion' => $this->getSignatureVersion($data),
+            'signingRegion' => isset($data['credentialScope']['region'])
+                ? $data['credentialScope']['region']
+                : $region,
+            'signingName' => isset($data['credentialScope']['service'])
+                ? $data['credentialScope']['service']
+                : $service,
+        ];
+    }
+
+    private function getEndpointData($service, $region)
+    {
+
+        $resolvedRegion = $this->resolveRegion($service, $region);
+        $data = isset($this->definition['services'][$service]['endpoints'][$resolvedRegion])
+            ? $this->definition['services'][$service]['endpoints'][$resolvedRegion]
+            : [];
+        $data += isset($this->definition['services'][$service]['defaults'])
+            ? $this->definition['services'][$service]['defaults']
+            : [];
+        $data += isset($this->definition['defaults'])
+            ? $this->definition['defaults']
+            : [];
+
+        return $data;
+    }
+
+    private function getSignatureVersion(array $data)
+    {
+        static $supportedBySdk = [
+            's3v4',
+            'v4',
+            'anonymous',
+        ];
+
+        $possibilities = array_intersect(
+            $supportedBySdk,
+            isset($data['signatureVersions'])
+                ? $data['signatureVersions']
+                : ['v4']
+        );
+
+        return array_pop($possibilities);
```

Comment:
I'm using `array_intersect` here to make sure that only signature versions that are supported by the SDK are pulled from the endpoints file. For example, Import/Export's signature versions are: `['v2', 'v4']`; the intersection of that against the SDK-supported signature versions will return an array with just one element: `v4`
```php
$supported = ['s3v4', 'v4', 'anonymous'];
$valid = array_intersect($supported, ['v2', 'v4']);
// $valid is [1 => 'v4']; 
```

If the service and SDK support more than one signature version, we should use them in the order of preference provided by the `$supportedBySdk` variable. I should be using `array_shift` instead of `array_pop` to enforce that.